### PR TITLE
Add precertificate generation to CA.

### PR DESCRIPTION
**Pull request by taknira**

### CODE REVIEW

File: certgen/certgen.go [hunk 1]
```diff
@@ -83,6 +83,43 @@ type CA struct {
 // in the SigningCert and SigningKey fields of the CA, and configured using the
 // CertConfig in the CA.
 func (ca *CA) IssueCertificate() (*x509.Certificate, error) {
+	return ca.issueCertificate(false)
+}
+
+// IssueCertificateChain creates a new certificate chain, containing a new leaf
```
Comment (phad) [inline certgen/certgen.go]: new twice; maybe ding the first? It's implicitly a new chain but it's worth saying the leaf itself is new.

File: certgen/certgen.go [hunk 2]
```diff
@@ -83,6 +83,43 @@ type CA struct {
...
+	}
+
+	return []*x509.Certificate{leaf, ca.SigningCert}, nil
+}
+
+// IssuePrecertificate creates a new leaf precertificate, issued by the key
+// specified in the SigningCert and SigningKey fields of the CA, and configured
+// using the CertConfig in the CA.
+func (ca *CA) IssuePrecertificate() (*x509.Certificate, error) {
+	return ca.issueCertificate(true)
```
Comment (phad) [inline certgen/certgen.go]: It's nice to comment the meaning of these bare bool vals, e.g. `issueCertificate(true /*precert*/)`

File: certgen/certgen.go [hunk 3]
```diff
@@ -83,6 +83,43 @@ type CA struct {
...
+}
+
+// IssuePrecertificateChain creates a new certificate chain, containing a new
```
Comment (phad) [inline certgen/certgen.go]: same repeated 'new' niggle here

File: certgen/certgen_test.go [hunk 4]
```diff
@@ -105,9 +126,15 @@ func TestIssueCertificate(t *testing.T) {
...
+			var cert *x509.Certificate
+			var err error
+			if test.precert {
+				cert, err = ca.IssuePrecertificate()
+			} else {
+				cert, err = ca.IssueCertificate()
+			}
 			if err != nil {
-				t.Fatalf("error creating certificate: %s", err)
+				t.Fatalf("error creating (pre)certificate: %s", err)
```
Comment (phad) [inline certgen/certgen_test.go]: If this fails the test you'll get a t.Run(desc) prefix
which hopefully identifies whether it was pre- or not.

But to be safe you might do something like:
`t.Fatalf("error creating (pre? %t)-certificate: %s", test.precert, err)`

File: certgen/certgen_test.go [hunk 5]
```diff
@@ -164,6 +191,9 @@ func TestIssueCertificate(t *testing.T) {
 			if cert.IsCA {
 				t.Errorf("certificate IsCA = %t, want false", cert.IsCA)
 			}
+			if test.precert && !cert.IsPrecertificate() {
```
Comment (phad) [inline certgen/certgen_test.go]: I think you're missing the converse in this test.
ie.  `!test.precert && cert.IsPrecertificate()`

Comment (phad) [reply certgen/certgen_test.go]: So you might need to do the pattern we see for `gotErr, wantErr`
i.e.

```
if gotPre, wantPre := cert.IsPrecertificate(), test.precert; gotPre != wantPre {
  t.Errorf("certificate type wrong: got precert? %t want precert? %t", gotPre, wantPre)
}
```

Comment (taknira) [reply certgen/certgen_test.go]: Done.

**Pull request merged**